core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MakeAsyncHttpRequestStage.java

```diff
@@ -136,119 +153,83 @@ private boolean shouldSetContentLength(SdkHttpFullRequest request, SdkHttpReques
 
     }
 
+    private void setupAttemptTimer(CompletableFuture<Response<OutputT>> executeFuture, RequestExecutionContext ctx) {
+        final long timeoutMillis = apiCallAttemptTimeoutInMillis(ctx.requestConfig());
+        TimerUtils.timeCompletableFuture(executeFuture,
+                                         timeoutExecutor,
+                                         ApiCallAttemptTimeoutException.create(timeoutMillis),
+                                         timeoutMillis);
+    }
+
     /**
      * Detects whether the response succeeded or failed and delegates to appropriate response handler.
      */
-    private class ResponseHandler implements SdkHttpResponseHandler<Response<OutputT>> {
-        private final Completable completable;
 
-        private volatile SdkHttpResponse response;
-        private volatile boolean isSuccess = false;
+    private class ResponseHandler implements TransformingAsyncResponseHandler<Response<OutputT>> {
+        private final CompletableFuture<SdkHttpResponse> headersFuture = new CompletableFuture<>();
+        private final CompletableFuture<OutputT> transformFuture;
+        private final CompletableFuture<? extends SdkException> errorTransformFuture;
+        private volatile SdkHttpFullResponse response;
 
-        /**
-         * @param completable   Future to notify when response has been handled.
-         */
-        private ResponseHandler(Completable completable) {
-            this.completable = completable;
+        ResponseHandler(CompletableFuture<OutputT> transformFuture,
+                        CompletableFuture<? extends SdkException> errorTransformFuture) {
+            this.transformFuture = transformFuture;
+            this.errorTransformFuture = errorTransformFuture;
         }
 
         @Override
-        public void headersReceived(SdkHttpResponse response) {
+        public void onHeaders(SdkHttpResponse response) {
+            headersFuture.complete(response);
             if (response.isSuccessful()) {
                 SdkStandardLogger.REQUEST_LOGGER.debug(() -> "Received successful response: " + response.statusCode());
-                isSuccess = true;
-                responseHandler.headersReceived(response);
+                responseHandler.onHeaders(response);
             } else {
                 SdkStandardLogger.REQUEST_LOGGER.debug(() -> "Received error response: " + response.statusCode());
-                errorResponseHandler.headersReceived(response);
+                errorResponseHandler.onHeaders(response);
+            }
+            try {
+                this.response = (SdkHttpFullResponse) response;
+            } catch (ClassCastException e) {
+                // Make SpotBugs happy. Should never happen since SdkHttpFullResponse is the only concrete impl we pass
+                // around.
+                log.error(() -> response + " is not an instance of SdkHttpFullResponse", e);
```

Comment:
Ew. If it's the only impl we'll pass around, why can't we make it typed correctly?

Reply:
I'm not sure, that's the way it was originally. I think we should be able to fix the types though!